zilliqa/src/contracts/utils/deque_v2.sol

```diff
@@ -0,0 +1,74 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity ^0.8.20;
+
+struct Withdrawal {
+    uint256 startedAt;
+    uint256 amount;
+}
+
+// A non-broken implementation of a double-ended queue of `Withdrawal`s. `deque.sol` doesn't work, don't use it.
+library Deque {
+    struct Withdrawals {
+        mapping(uint256 index => Withdrawal) values;
+        // The index of the first element.
+        uint256 head;
+        // The index where the next element will be inserted at the end of the queue.
+        uint256 tail;
+    }
+
+    function length(Withdrawals storage deque) internal view returns (uint256) {
+        return deque.tail - deque.head;
+    }
+
+    // Get the element at the given logical index. Reverts if `idx >= queue.length()`.
+    function get(
+        Withdrawals storage deque,
+        uint256 idx
+    ) internal view returns (Withdrawal storage) {
+        require(deque.head + idx < deque.tail, "element does not exist");
+        return deque.values[deque.head + idx];
+    }
+
+    // Push an empty element to the back of the queue. Returns a reference to the new element.
+    function pushBack(
+        Withdrawals storage deque
+    ) internal returns (Withdrawal storage) {
+        uint256 idx = deque.tail;
+        deque.tail++;
+        return deque.values[idx];
+    }
+
+    // Pop an element from the front of the queue.
+    function popFront(
+        Withdrawals storage deque
+    ) internal returns (Withdrawal memory) {
+        require(deque.head < deque.tail, "queue is empty");
```

Comment:
I  think this should be `<=` ?

Reply:
If `head == tail`, the queue is empty. So we require that `head != tail` (i.e. `head < tail`)

Reply:
Sorry!